### PR TITLE
[DSM] DDP-6584 Downloading participant files

### PR DESCRIPTION
**Pull request by pegahtah**
I have added a new component `file-download` and a new tab in participant page to view the uploaded files. 
Also new columns for participant list has been added to enable searching on the fields
User must have permission and the participant must have uploaded files for the tab to appear.
Screen Shot 2022-07-26 at 11 50 32 AM

Screen Shot 2022-07-25 at 2 01 41 PM

### CODE REVIEW

File: ddp-workspace/projects/ddp-dsm-ui/src/app/file-download/file-download.component.html [hunk 1]
```diff
@@ -0,0 +1,38 @@
+<ng-container>
+  <ng-container *ngIf="downloading">
+    <div align="center">
```
Comment (alextanlan) [inline ddp-workspace/projects/ddp-dsm-ui/src/app/file-download/file-download.component.html]: `align` attribute is obsolete (see [link](https://developer.mozilla.org/en-US/docs/web/html/element/div#attr-align:~:text=Note%3A%20The%20align%20attribute%20is%20obsolete%3B%20do%20not%20use%20it%20anymore.%20Instead%2C%20you%20should%20use%20CSS%20properties%20or%20techniques%20such%20as%20CSS%20Grid%20or%20CSS%20Flexbox%20to%20align%20and%20position%20%3Cdiv%3E%20elements%20on%20the%20page.)).
Use css instead (text-align: center) .

File: ddp-workspace/projects/ddp-dsm-ui/src/app/file-download/file-download.component.html [hunk 2]
```diff
@@ -0,0 +1,38 @@
+<ng-container>
```
Comment (alextanlan) [inline ddp-workspace/projects/ddp-dsm-ui/src/app/file-download/file-download.component.html]: Wondering why there are so many `ng-container`-s here? 
Yu can get rid of all of them, like:
image

File: ddp-workspace/projects/ddp-dsm-ui/src/app/file-download/file-download.component.html [hunk 3]
```diff
@@ -0,0 +1,38 @@
...
+      </tr>
+      </thead>
+      <tr *ngFor="let file of participant.data.files">
+        <td>{{file.fileName}}</td>
+        <td>{{getNiceDateFormat(file.uploadedAt)}}</td>
+        <td>{{file.scanResult}}</td>
+        <td>
+          <button (click)="downloadParticipantFile(file)" [disabled]="file.scanResult !== 'CLEAN'"
+                  color="primary" mat-raised-button
+                  type="button">Download File
```
Comment (alextanlan) [inline ddp-workspace/projects/ddp-dsm-ui/src/app/file-download/file-download.component.html]: I believe `type=button` can be skipped.

File: ddp-workspace/projects/ddp-dsm-ui/src/app/file-download/file-download.component.ts [hunk 4]
```diff
@@ -0,0 +1,66 @@
...
+    const url = window.URL.createObjectURL( blob );
+    window.open( url );
+
+    const shortId = this.participant.data.profile[ 'hruid' ];
+
+    fileSaver.saveAs( blob, shortId + '_' + fileName );
+  }
+
+
+  getNiceDateFormat( uploadedAt: any ): string {
```
Comment (alextanlan) [inline ddp-workspace/projects/ddp-dsm-ui/src/app/file-download/file-download.component.ts]: `uploadedAt` should be a string according to `ddp-dsm-ui\src\app\participant-list\models\file.model.ts`.
Is not it?

File: ddp-workspace/projects/ddp-dsm-ui/src/app/file-download/file-download.component.html [hunk 5]
```diff
@@ -0,0 +1,38 @@
...
+        </th>
+        <th>
+          Time of Upload
+        </th>
+        <th colspan="2">
+          Scan Result
+        </th>
+      </tr>
+      </thead>
+      <tr *ngFor="let file of participant.data.files">
```
Comment (alextanlan) [inline ddp-workspace/projects/ddp-dsm-ui/src/app/file-download/file-download.component.html]: What if `participant` or `participant.data` are null/undefined ?

File: ddp-workspace/projects/ddp-dsm-ui/src/app/file-download/file-download.component.ts [hunk 6]
```diff
@@ -0,0 +1,66 @@
...
+export class FileDownloadComponent {
+  @Input() participant: Participant;
+  downloadMessage = '';
+  downloading = false;
+
+  constructor( private role: RoleService, private dsmService: DSMService ) {
+
+  }
+
+  hasRole(): RoleService {
```
Comment (alextanlan) [inline ddp-workspace/projects/ddp-dsm-ui/src/app/file-download/file-download.component.ts]: the method `hasRole` is nowhere used (as well as `RoleService` itself).

File: ddp-workspace/projects/ddp-dsm-ui/src/app/file-download/file-download.component.html [hunk 7]
```diff
@@ -0,0 +1,38 @@
...
+        <td>{{file.fileName}}</td>
+        <td>{{getNiceDateFormat(file.uploadedAt)}}</td>
+        <td>{{file.scanResult}}</td>
+        <td>
+          <button (click)="downloadParticipantFile(file)" [disabled]="file.scanResult !== 'CLEAN'"
```
Comment (alextanlan) [inline ddp-workspace/projects/ddp-dsm-ui/src/app/file-download/file-download.component.html]: It makes sense to keep 'CLEAN' string as a constant.

File: ddp-workspace/projects/ddp-dsm-ui/src/app/file-download/file-download.component.ts [hunk 8]
```diff
@@ -0,0 +1,66 @@
...
+
+  downloadParticipantFile( file: File ): void {
+    this.downloading = true;
+    if (file.scanResult !== 'CLEAN') {
+      this.downloadMessage = 'Error - file has not passed scanning';
+      this.downloading = false;
+      return;
+    }
+    const realm = localStorage.getItem( ComponentService.MENU_SELECTED_REALM );
+    this.dsmService.downloadParticipantFile( file.fileName, file.bucket, file.blobName, realm, file.mimeType ).subscribe(
```
Comment (alextanlan) [inline ddp-workspace/projects/ddp-dsm-ui/src/app/file-download/file-download.component.ts]: Can you please add `unsubscribe` for the subscription?

File: ddp-workspace/projects/ddp-dsm-ui/src/app/file-download/file-download.component.ts [hunk 9]
```diff
@@ -0,0 +1,66 @@
...
+        console.log( err );
+        this.downloadMessage = err;
+        this.downloading = false;
+      }
+    );
+  }
+
+  saveParticipantFile( data: any, type: string, fileName: string ): void {
+    const blob = new Blob( [ data ], {type: type} );
+    const url = window.URL.createObjectURL( blob );
```
Comment (alextanlan) [inline ddp-workspace/projects/ddp-dsm-ui/src/app/file-download/file-download.component.ts]: Just wondering what are lines 54-55 for?

File: ddp-workspace/projects/ddp-dsm-ui/src/app/participant-list/models/file.model.ts [hunk 10]
```diff
@@ -0,0 +1,18 @@
...
+
+  constructor( public bucket: string, public uploadedAt: string, public fileName: string, public mimeType: string,
+               public scanResult: string, public fileSize: number, public blobName: string ) {
+    this.bucket = bucket;
+    this.uploadedAt = uploadedAt;
+    this.fileName = fileName;
+    this.mimeType = mimeType;
+    this.scanResult = scanResult;
+    this.fileSize = fileSize;
+    this.blobName = blobName;
```
Comment (alextanlan) [inline ddp-workspace/projects/ddp-dsm-ui/src/app/participant-list/models/file.model.ts]: Lines 5-11 can be easily removed.

File: ddp-workspace/projects/ddp-dsm-ui/src/app/participant-list/models/data.model.ts [hunk 11]
```diff
@@ -21,6 +23,7 @@ export class Data {
     this.address = address;
     this.invitations = invitations;
     this.computed = computed;
+    this.files = files;
```
Comment (alextanlan) [inline ddp-workspace/projects/ddp-dsm-ui/src/app/participant-list/models/data.model.ts]: Lines 16-26 are not needed, can be deleted. 
Typescript does that job under the hood.

File: ddp-workspace/projects/ddp-dsm-ui/src/app/participant-list/participant-list.component.html [hunk 12]
```diff
@@ -541,6 +541,18 @@ <h3 class="Color--primary Line--Break">{{additionalMessage}}</h3>
                     </ng-container>
                   </div>
 
+                  <div *ngIf="col.participantColumn.tableAlias==='data' && col.participantColumn.object==='files'">
+                    <ul class="NO-PADDING-START">
+                      <li class="LIST-SEPARATOR" *ngFor="let file of pt.data.files">
+                        <p class="BIT-OF-MARGIN">
+                          <ng-container>
```
Comment (alextanlan) [inline ddp-workspace/projects/ddp-dsm-ui/src/app/participant-list/participant-list.component.html]: I can not see any benefits for using ng-container here.

File: ddp-workspace/projects/ddp-dsm-ui/src/app/services/dsm.service.ts [hunk 13]
```diff
@@ -611,6 +611,17 @@ export class DSMService {
...
   }
 
+  public downloadParticipantFile( fileName: string, bucketName: string, blob: string, realm: string, mimeType: string ): Observable<any> {
+    const url = this.baseUrl + DSMService.UI + 'downloadFile';
+    const map: { name: string; value: any }[] = [];
+    map.push( {name: DSMService.REALM, value: realm} );
+    map.push( {name: 'fileName', value: fileName} );
+    map.push( {name: 'bucket', value: bucketName} );
+    map.push( {name: 'blob', value: blob} );
+    map.push( {name: 'responseType', value: 'blob'} );
```
Comment (alextanlan) [inline ddp-workspace/projects/ddp-dsm-ui/src/app/services/dsm.service.ts]: **_#optional_**
The line is unnecessary. 
That parameter is already in `buildQueryBlobHeader` method.

**Comment by alextanlan** [comment]
Can you please fix linter warnings?
image

File: ddp-workspace/projects/ddp-dsm-ui/src/app/file-download/file-download.component.ts (hunk 8 above)
Comment (pegahtah) [reply ddp-workspace/projects/ddp-dsm-ui/src/app/file-download/file-download.component.ts]: Thanks! I wasn't aware we have to unsubscribe as well

File: ddp-workspace/projects/ddp-dsm-ui/src/app/participant-list/participant-list.component.html (hunk 12 above)
Comment (pegahtah) [reply ddp-workspace/projects/ddp-dsm-ui/src/app/participant-list/participant-list.component.html]: As it seems, I unknowingly tend to over use them. Thanks for catching that!

File: ddp-workspace/projects/ddp-dsm-ui/src/app/file-download/file-download.component.ts (hunk 8 above)
Comment (alextanlan) [reply ddp-workspace/projects/ddp-dsm-ui/src/app/file-download/file-download.component.ts]: https://blog.bitsrc.io/6-ways-to-unsubscribe-from-observables-in-angular-ab912819a78f

File: ddp-workspace/projects/ddp-dsm-ui/src/app/participant-list/participant-list.component.html (hunk 12 above)
Comment (alextanlan) [reply ddp-workspace/projects/ddp-dsm-ui/src/app/participant-list/participant-list.component.html]: Often it is used when you have more than 1 structural directive on an element, e.g. ***ngIf** and ***ngFor**
see [examples](https://www.digitalocean.com/community/tutorials/angular-ng-container-element )

**Review by alextanlan — APPROVED** [review]

**Pull request merged**